File: avalon/composition/impl/src/java/org/apache/avalon/composition/model/impl/DefaultDependencyModel.java
/* 
 * Copyright 2004 Apache Software Foundation
 * Licensed  under the  Apache License,  Version 2.0  (the "License");
 * you may not use  this file  except in  compliance with the License.
 * You may obtain a copy of the License at 
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed  under the  License is distributed on an "AS IS" BASIS,
 * WITHOUT  WARRANTIES OR CONDITIONS  OF ANY KIND, either  express  or
 * implied.
 * 
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.avalon.composition.model.impl;

import java.util.ArrayList;

import org.apache.avalon.composition.model.DependencyModel;
import org.apache.avalon.composition.model.ModelException;
import org.apache.avalon.composition.data.DependencyDirective;
import org.apache.avalon.composition.data.SelectionDirective;

import org.apache.avalon.framework.logger.Logger;

import org.apache.avalon.meta.info.DependencyDescriptor;
import org.apache.avalon.meta.info.ServiceDescriptor;

import org.apache.avalon.util.i18n.ResourceManager;
import org.apache.avalon.util.i18n.Resources;

/**
 * Default implementation of the deplendency model.
 *
 * @author <a href="mailto:dev@avalon.apache.org">Avalon Development Team</a>
 * @version $Revision: 1.9 $ $Date: 2004/03/17 10:39:10 $
 */
public class DefaultDependencyModel extends DefaultDependent implements DependencyModel
{
    //--------------------------------------------------------------
    // static
    //--------------------------------------------------------------

    private static final Resources REZ =
            ResourceManager.getPackageResources( DefaultDependencyModel.class );

    //--------------------------------------------------------------
    // immutable state
    //--------------------------------------------------------------

    private final DependencyDescriptor m_descriptor;

    private final DependencyDirective m_directive;

    private final String m_partition;

    private final String m_name;

    private final String m_source;

    //--------------------------------------------------------------
    // constructor
    //--------------------------------------------------------------

   /**
    * Creation of a new dependency model.
    *
    * @param logger the logging channel
    * @param partition the partition
    * @param name the name
    * @param descriptor the dependency descriptor
    * @param directive the dependency directive (possibly null)
    */
    public DefaultDependencyModel( 
      final Logger logger, final String partition, final String name, 
      final DependencyDescriptor descriptor, DependencyDirective directive )
      throws ModelException
    {
        super( logger );

        if( descriptor == null ) throw new NullPointerException( "descriptor" );

        m_descriptor = descriptor;
        m_directive = directive;
        m_partition = partition;
        m_name = name;

        //
        // a dependency directive is either declaring an explicitly
        // identified provider, or, it is delcaring 0 or more selection 
        // constraints - if its a an absolute source declaration then 
        // add it to the table to paths keyed by depedency key names
        //

        if( directive != null )
        {
            if( directive.getSource() != null )
            {
                m_source = resolvePath( partition, directive.getSource() );
                final String message =
                  REZ.getString( "dependency.path.debug", m_source, directive.getKey() );
                getLogger().debug( message );
            }
            else
            {
                m_source = null;
            }
        }
        else
        {
            m_source = null;
        }
    }

    //--------------------------------------------------------------
    // DependencyModel
    //--------------------------------------------------------------

   /**
    * Return the dependency descriptor.
    *
    * @return the descriptor
    */
    public DependencyDescriptor getDependency()
    {
        return m_descriptor;
    }

   /**
    * Return an explicit path to a supplier component.  
    * If a dependency directive has been declared
    * and the directive contains a source declaration, the value 
    * returned is the result of parsing the source value relative 
    * to the absolute address of the implementing component.
    *
    * @return the explicit path
    */
    public String getPath()
    {
        return m_source;
    }

   /**
    * Filter a set of candidate service descriptors and return the 
    * set of acceptable service as a ordered sequence.
    *
    * @param candidates the set of candidate services for the dependency
    *    matching the supplied key
    * @return the accepted candidates in ranked order
    * @exception IllegalArgumentException if the key is unknown
    */
    public ServiceDescriptor[] filter( ServiceDescriptor[] candidates ) 
    {
        if( m_directive != null )
        {
            if( m_directive.getSource() == null )
            {
                return filter( m_directive, candidates );
            }
        }
        return candidates;
    }

   /**
    * Filter a set of candidate service descriptors and return the 
    * set of acceptable service as a ordered sequence.
    *
    * @param directive the dependency directive
    * @param services the set of candidate services for the dependency
    * @return the accepted candidates in ranked order
    */
    private ServiceDescriptor[] filter( 
      DependencyDirective directive, ServiceDescriptor[] services ) 
    {
        SelectionDirective[] filters = getFilters( directive );
        ArrayList list = new ArrayList();

        for( int i=0; i<services.length; i++ )
        {
            ServiceDescriptor service = services[i];
            if( isaCandidate( service, filters ) )
            {
                list.add( service );
            }
        }

        ServiceDescriptor[] candidates = 
          (ServiceDescriptor[]) list.toArray( new ServiceDescriptor[0] );

        //
        // TODO: include ranking of candidates
        //

        return candidates;
    }

    private boolean isaCandidate( 
      ServiceDescriptor service, SelectionDirective[] filters )
    {
        for( int i=0; i<filters.length; i++ )
        {
            SelectionDirective filter = filters[i];
            if( !isaCandidate( service, filter ) )
            {
                return false;
            }
        }
        return true;
    }

    private boolean isaCandidate( 
      ServiceDescriptor service, SelectionDirective filter )
    {
        final String feature = filter.getFeature();
        final String value = filter.getValue();
        final String criteria = filter.getCriteria();

        if( criteria.equals( SelectionDirective.EQUALS ) )
        {
            return value.equals( service.getAttribute( feature ) );
        }
        else if( criteria.equals( SelectionDirective.EXISTS ) )
        {
            return service.getAttribute( feature ) != null;
        }
        else if( criteria.equals( SelectionDirective.INCLUDES ) )
        {
            final String v = service.getAttribute( feature );
            if( v != null )
            {
                return v.indexOf( value ) > -1;
            }
            else
            {
                return false;
            }
        }
        else
        {
            final String error = 
              REZ.getString( "dependency.invalid-criteria.error", criteria, feature );
            throw new IllegalArgumentException( error );
        }
    }

    private String resolvePath( String partition, String path )
    {
        if( path.startsWith( "/" ) )
        {
            return path;
        }
        else if( path.startsWith( "../" ) )
        {
            final String parent = getParentPath( partition );
            return resolvePath( parent, path.substring( 3 ) );
        }
        else if( path.startsWith( "./" ) )
        {
            return resolvePath( partition, path.substring( 2 ) );
        }
        else
        {
            return partition + path;
        }
    }

    private String getParentPath( String partition )
    {
        int n = partition.lastIndexOf( "/" );
        if( n > 0 )
        {
            int index = partition.substring( 0, n-1 ).lastIndexOf( "/" );
            if( index == 0 )
            {
                return "/";
            }
            else
            {
                return partition.substring( 0, index ) + "/";
            }
        }
        else
        {
            final String error = 
              "Illegal attempt to reference a containment context above the root context.";
            throw new IllegalArgumentException( error );
        }
    }

   /**
    * Return the required selection constraints.
    * @param directive the dependency directive
    * @return the set of required selection directives
    */
    private SelectionDirective[] getFilters( DependencyDirective directive )
    {
        ArrayList list = new ArrayList();
        SelectionDirective[] selections = directive.getSelectionDirectives();
        for( int i=0; i<selections.length; i++ )
        {
            SelectionDirective selection = selections[i];
            if( selection.isRequired() )
            {
                list.add( selection );
            }
        }
        return (SelectionDirective[]) list.toArray( new SelectionDirective[0] );
    }
}

File: legacy/avalon-sandbox/avalon-net/Composition/ModelDefault/DefaultDependencyModel.cs
/* 
* Copyright 2003-2004 The Apache Software Foundation
* Licensed  under the  Apache License,  Version 2.0  (the "License");
* you may not use  this file  except in  compliance with the License.
* You may obtain a copy of the License at 
* 
*   http://www.apache.org/licenses/LICENSE-2.0
* 
* Unless required by applicable law or agreed to in writing, software
* distributed  under the  License is distributed on an "AS IS" BASIS,
* WITHOUT  WARRANTIES OR CONDITIONS  OF ANY KIND, either  express  or
* implied.
* 
* See the License for the specific language governing permissions and
* limitations under the License.
*/
namespace Apache.Avalon.Composition.Model.Default
{
	using System;

	using Apache.Avalon.Framework;
	using Apache.Avalon.Meta;
	using Apache.Avalon.Composition.Model;
	using Apache.Avalon.Composition.Data;
	
	/// <summary> Default implementation of the deplendency model.
	/// 
	/// </summary>
	/// <author>  <a href="mailto:dev@avalon.apache.org">Avalon Development Team</a>
	/// </author>
	/// <version>  $Revision: 1.3 $ $Date: 2004/03/07 22:06:40 $
	/// </version>
	public class DefaultDependencyModel : DefaultDependent, IDependencyModel
	{
		//--------------------------------------------------------------
		// static
		//--------------------------------------------------------------
		
		//--------------------------------------------------------------
		// immutable state
		//--------------------------------------------------------------
		
		private DependencyDescriptor m_descriptor;
		
		private DependencyDirective m_directive;
		
		private String m_partition;
		
		private String m_name;
		
		private String m_source;
		
		//--------------------------------------------------------------
		// constructor
		//--------------------------------------------------------------
		
		/// <summary> Creation of a new dependency model.
		/// 
		/// </summary>
		/// <param name="logger">the logging channel
		/// </param>
		/// <param name="partition">the partition
		/// </param>
		/// <param name="name">the name
		/// </param>
		/// <param name="descriptor">the dependency descriptor
		/// </param>
		/// <param name="directive">the dependency directive (possibly null)
		/// </param>
		public DefaultDependencyModel(ILogger logger, String partition, String name, DependencyDescriptor descriptor, DependencyDirective directive):base(logger)
		{
			
			if (descriptor == null)
				throw new System.ArgumentNullException("descriptor");
			
			m_descriptor = descriptor;
			m_directive = directive;
			m_partition = partition;
			m_name = name;
			
			//
			// a dependency directive is either declaring an explicitly
			// identified provider, or, it is delcaring 0 or more selection 
			// constraints - if its a an absolute source declaration then 
			// add it to the table to paths keyed by depedency key names
			//
			
			if (directive != null)
			{
				if ((System.Object) directive.Source != null)
				{
					m_source = resolvePath(partition, directive.Source);
					String message = "dependency.path.debug" + " " + m_source + " " + directive.Key;
					Logger.Debug(message);
				}
				else
				{
					m_source = null;
				}
			}
			else
			{
				m_source = null;
			}
		}
		
		//--------------------------------------------------------------
		// IDependencyModel
		//--------------------------------------------------------------
		
		/// <summary> Return the dependency descriptor.
		/// 
		/// </summary>
		/// <returns> the descriptor
		/// </returns>
		public virtual DependencyDescriptor Dependency
		{
			get
			{
				return m_descriptor;
			}
		}
		
		/// <summary> Return an explicit path to a supplier component.  
		/// If a dependency directive has been declared
		/// and the directive contains a source declaration, the value 
		/// returned is the result of parsing the source value relative 
		/// to the absolute address of the implementing component.
		/// </summary>
		/// <returns> the explicit path
		/// </returns>
		public virtual String Path
		{
			get
			{
				return m_source;
			}
		}
		
		/// <summary> Filter a set of candidate service descriptors and return the 
		/// set of acceptable service as a ordered sequence.
		/// 
		/// </summary>
		/// <param name="candidates">the set of candidate services for the dependency
		/// matching the supplied key
		/// </param>
		/// <returns> the accepted candidates in ranked order
		/// </returns>
		/// <exception cref=""> IllegalArgumentException if the key is unknown
		/// </exception>
		public virtual ServiceDescriptor[] Filter(ServiceDescriptor[] candidates)
		{
			if (m_directive != null)
			{
				if ((System.Object) m_directive.Source == null)
				{
					return Filter(m_directive, candidates);
				}
			}
			return candidates;
		}
		
		/// <summary> Filter a set of candidate service descriptors and return the 
		/// set of acceptable service as a ordered sequence.
		/// 
		/// </summary>
		/// <param name="directive">the dependency directive
		/// </param>
		/// <param name="services">the set of candidate services for the dependency
		/// </param>
		/// <returns> the accepted candidates in ranked order
		/// </returns>
		private ServiceDescriptor[] Filter(DependencyDirective directive, ServiceDescriptor[] services)
		{
			SelectionDirective[] filters = GetFilters(directive);
			System.Collections.ArrayList list = new System.Collections.ArrayList();
			
			for (int i = 0; i < services.Length; i++)
			{
				ServiceDescriptor service = services[i];
				if (IsaCandidate(service, filters))
				{
					list.Add(service);
				}
			}
			
			ServiceDescriptor[] candidates = (ServiceDescriptor[]) list.ToArray( typeof(ServiceDescriptor) );
			
			//
			// TODO: include ranking of candidates
			//
			
			return candidates;
		}
		
		private bool IsaCandidate(ServiceDescriptor service, SelectionDirective[] filters)
		{
			for (int i = 0; i < filters.Length; i++)
			{
				SelectionDirective filter = filters[i];
				if (!IsaCandidate(service, filter))
				{
					return false;
				}
			}
			return true;
		}
		
		private bool IsaCandidate(ServiceDescriptor service, SelectionDirective filter)
		{
			String feature = filter.Feature;
			String value_Renamed = filter.Value;
			String criteria = filter.Criteria;
			
			if (criteria.Equals(SelectionDirective.EQUALS))
			{
				return value_Renamed.Equals(service.GetAttribute(feature));
			}
			else if (criteria.Equals(SelectionDirective.EXISTS))
			{
				return service.GetAttribute(feature) != null;
			}
			else if (criteria.Equals(SelectionDirective.INCLUDES))
			{
				String v = service.GetAttribute(feature);
				if ((System.Object) v != null)
				{
					return v.IndexOf(value_Renamed) > - 1;
				}
				else
				{
					return false;
				}
			}
			else
			{
				String error = "dependency.invalid-criteria.error" + " " + criteria + " " + feature;
				throw new System.ArgumentException(error);
			}
		}
		
		private String resolvePath(String partition, String path)
		{
			if (path.StartsWith("/"))
			{
				return path;
			}
			else if (path.StartsWith("../"))
			{
				String parent = GetParenPath(partition);
				return resolvePath(parent, path.Substring(3));
			}
			else if (path.StartsWith("./"))
			{
				return resolvePath(partition, path.Substring(2));
			}
			else
			{
				return partition + path;
			}
		}
		
		private String GetParenPath(String partition)
		{
			int n = partition.LastIndexOf("/");
			if (n > 0)
			{
				int index = partition.Substring(0, (n - 1) - (0)).LastIndexOf("/");
				if (index == 0)
				{
					return "/";
				}
				else
				{
					return partition.Substring(0, (index) - (0)) + "/";
				}
			}
			else
			{
				String error = "Illegal attempt to reference a containment context above the root context.";
				throw new System.ArgumentException(error);
			}
		}
		
		/// <summary> Return the required selection constraints.</summary>
		/// <param name="directive">the dependency directive
		/// </param>
		/// <returns> the set of required selection directives
		/// </returns>
		private SelectionDirective[] GetFilters(DependencyDirective directive)
		{
			System.Collections.ArrayList list = new System.Collections.ArrayList();
			SelectionDirective[] selections = directive.SelectionDirectives;
			for (int i = 0; i < selections.Length; i++)
			{
				SelectionDirective selection = selections[i];
				if (selection.Required)
				{
					list.Add(selection);
				}
			}
			return (SelectionDirective[]) list.ToArray( typeof(SelectionDirective) );
		}
	}
}